sdks/java/testing/test-utils/src/main/java/org/apache/beam/sdk/testutils/metrics/MetricsReader.java

```diff
@@ -0,0 +1,126 @@
+
+  private final String namespace;
+
+  public MetricsReader(PipelineResult result, String namespace) {
+    this.result = result;
+    this.namespace = namespace;
+  }
+
+  /**
+   * Return the current value for a long counter, or a default value if can't be retrieved. Note
+   * this uses only attempted metrics because some runners don't support committed metrics.
+   */
+  public long getCounterMetric(String name, long defaultValue) {
+    MetricQueryResults metrics =
+        result
+            .metrics()
+            .queryMetrics(
+                MetricsFilter.builder()
+                    .addNameFilter(MetricNameFilter.named(namespace, name))
+                    .build());
+    Iterable<MetricResult<Long>> counters = metrics.getCounters();
+    try {
+      MetricResult<Long> metricResult = counters.iterator().next();
+      return metricResult.getAttempted();
+    } catch (NoSuchElementException e) {
+      LOG.error("Failed to get metric {}, from namespace {}", name, namespace);
+    }
+    return defaultValue;
+  }
+
+  /**
+   * Return start time metric by counting the difference between "now" and min value from a
+   * distribution metric.
+   */
+  public long getStartTimeMetric(long now, String name) {
+    return this.getTimestampMetric(now, this.getDistributionMetric(name, DistributionType.MIN, -1));
+  }
+
+  /**
+   * Return end time metric by counting the difference between "now" and MAX value from a
+   * distribution metric.
+   */
+  public long getEndTimeMetric(long now, String name) {
+    return this.getTimestampMetric(now, this.getDistributionMetric(name, DistributionType.MAX, -1));
+  }
+
+  /**
+   * Return the current value for a long counter, or a default value if can't be retrieved. Note
+   * this uses only attempted metrics because some runners don't support committed metrics.
+   */
+  private long getDistributionMetric(String name, DistributionType distType, long defaultValue) {
+    MetricQueryResults metrics =
+        result
+            .metrics()
+            .queryMetrics(
+                MetricsFilter.builder()
+                    .addNameFilter(MetricNameFilter.named(namespace, name))
+                    .build());
+    Iterable<MetricResult<DistributionResult>> distributions = metrics.getDistributions();
+    try {
+      MetricResult<DistributionResult> distributionResult = distributions.iterator().next();
+      switch (distType) {
+        case MIN:
+          return distributionResult.getAttempted().getMin();
+        case MAX:
+          return distributionResult.getAttempted().getMax();
+        default:
+          return defaultValue;
+      }
+    } catch (NoSuchElementException e) {
+      LOG.error("Failed to get distribution metric {} for namespace {}", name, namespace);
+    }
+    return defaultValue;
+  }
+
+  /** Return the current value for a time counter, or -1 if can't be retrieved. */
+  private long getTimestampMetric(long now, long value) {
```

Comment:
I only refactored this but are not sure if this is the best way to check if the value for timestamp is valid. Probably the days count could be smaller, eg. 1 day, wdyt?

Reply:
For me, it's not clear why it's exactly `10000`. Does Metrics API in case of errors returns dates above 10000? Maybe original author knows anything about it?

About 1 day. Probably there can be jobs that run longer than 1 day.

Reply:
I believe it was just a large enough number to exclude most invalid values, most of our tests that use 1000+ cores are unlikely to run for more then 1 day so the summed time is unlikely to pass 10000 cumulative days of time.